File: Java/TP3_MET_JAVA/src/Iteratif/CaractereMotIteratif.java
package Iteratif;

import Interface.Mot;

public class CaractereMotIteratif implements Mot {

    public static final Mot FAB = MotVideIteratif.SINGLETON;

    private Character caractere;
    private Mot motDroite;

    public CaractereMotIteratif(Character caractere, Mot motDroite) {
        this.caractere = caractere;
        this.motDroite = motDroite;
    }

    public Character caractere() {
        return caractere;
    }

    public Mot getMotDroite() {
        return motDroite;
    }

    @Override
    public Mot reste() {
        return motDroite;
    }

    @Override
    public Mot creerMotVide() {
        return MotVideIteratif.SINGLETON;
    }

    @Override
    public Mot creerCaractereMot(char c) {
        return new CaractereMotIteratif(c, this);
    }

    @Override
    public Mot creerMotMot(Mot droite) {
        return new MotMotIteratif(this, droite);
    }

    public boolean casCaractereMot() {
        return true;
    }

}

File: TypeScript/TP3_MET_TS/src/Iteratif/CaractereMotIteratif.ts

import {MotMotIteratif} from "./MotMotIteratif";
import {MotVideIteratif} from "./MotVideIteratif";
import {Mot} from "../Interface/Mot";

export class CaractereMotIteratif extends Mot {

    public static FAB: Mot = MotVideIteratif.SINGLETON;

    private _caractere: string;
    private motDroite: Mot;

    public constructor(caractere: string, motDroite: Mot) {
        super();
        this._caractere = caractere;
        this.motDroite = motDroite;
    }

    public caractere(): string {
        return this._caractere;
    }

    public getMotDroite(): Mot {
        return this.motDroite;
    }

    public reste(): Mot {
        return this.motDroite;
    }

    public creerMotVide(): Mot {
        return MotVideIteratif.SINGLETON;
    }

    public creerCaractereMot(c: string): Mot {
        return new CaractereMotIteratif(c, this);
    }

    public creerMotMot(droite: Mot): Mot {
        return new MotMotIteratif(this, droite);
    }

    public casCaractereMot(): boolean {
        return true;
    }
}
